### ISSUE TITLE
[RFC] Introduce automatic terminator for hyperparameter optimization (Optuna Terminator 🤖)

### ISSUE THREAD

**Issue by g-votte** [optuna/optuna#4398]
### Motivation

This issue proposes a set of new features, referred to as "Optuna Terminator," for automatically stopping an Optuna study. Optuna Terminator is based on the method proposed by Makarova et al. in "[Automatic Termination for Hyperparameter Optimization](https://www.amazon.science/publications/automatic-termination-for-hyperparameter-optimization)" [1].

As the optimization progresses, the objective function often becomes noisy and the room left for the optimization decreases in many use cases. However, it can be difficult for the user to detect the noise and to determine when to terminate the optimization. The user tends to continue the study for a long time, even when the optimization is no longer beneficial, resulting in wasted computation resources and increased risk of overfitting the hyperparameters.

Optuna Terminator addresses this problem by visualizing the room left for optimization and automatically terminating the optimization at the appropriate time.

![image](https://user-images.githubusercontent.com/1061922/217133233-269e7d36-0fb6-4b8e-911e-9a51e3419fa8.png)

### Description

Optuna Terminator provides the following features:
- evaluating and visualizing the room left for the optimization as a metric called "regret bounds," and
- specifically with cross validation (CV), a callback to automatically stops the study when the CV variance exceeds the regret bound.

### Visualizing regret bounds
The following snippet provides an example of evaluating the regret bounds, i.e., the room left for the optimization.

```python
from optuna.visualization import plot_regret_bounds
...
study.optimize(objective, n_trials=50)
plot_regret_bounds(study).show()
```
Then, the room left for the optimization is visualized as shown in the following figure.

![image](https://user-images.githubusercontent.com/1061922/217133315-ae4944a1-d089-47d1-bfe7-23893c08cb0c.png)

### Automatic termination
The following snippet provides an example of automatically terminating the optimization.

```python
from optuna.terminator.callback import TerminatorCallback
...

def objective(trial):
    ...
    trial.set_user_attr("cv_scores", scores)
    ...

terminator = TerminatorCallback()
...
study.optimize(objective, callbacks=[terminator])
```

Then, the `study.stop` method will be automatically invoked when the CV variance exceeds the regret bound. Note that the number of trials, `n_trials`, is not specified in this example.

### References
[1] Makarova et al. "Automatic termination for hyperparameter optimization." International Conference on Automated Machine Learning (2022).

**Comment by g-votte** [comment]
Soon, I will create some draft PRs that contain prototype implementations.
